### ISSUE TITLE
Flow direction heuristic breaks on long-running UDP flows (and other scenarios)

### ISSUE THREAD

**Issue by fako1024** [els0r/goProbe#38]
TL;DR: The direction detection heuristic stumbles on long-running UDP connections (e.g. VPN connections), causing significant inconsistencies / errors in measured ingress / egress traffic.

Longer story:
While implementing AF_PACKET in https://github.com/els0r/goProbe/issues/34 I stumbled upon some flow mismatches when comparing the PCAP approach (by simply running two instances in parallel, one using PCAP, one using AF_PACKET). At first I thought it was related to, well, AF_PACKET, but it turns out that this can be reproduced by simply running two standard instances of `goProbe`. In the beginning, the output from goQuery on the two databases is identical, but after a few hours of running on my local device I obtained these results from the two DBs, respectively:

```
└─ $ ▶ goQuery -i any  -d /tmp/goprobe_db_pcap -f -10000d dport,proto -n 10

                              packets   packets             bytes      bytes       
        iface  dport  proto        in       out      %         in        out      %
  enp45s0u1u1    443    UDP    2.32 M  265.91 k  37.35    2.76 GB   25.99 MB  49.41
  enp45s0u1u1   3478    UDP  949.59 k    1.34 M  33.15  725.17 MB  871.80 MB  27.67
  enp45s0u1u1  50024    UDP  579.44 k  777.59 k  19.62  435.55 MB  477.46 MB  15.82
  enp45s0u1u1    443    TCP  114.95 k   96.88 k   3.06  192.06 MB   47.86 MB   4.16
  enp45s0u1u1   4500    UDP   89.55 k   63.22 k   2.21   58.12 MB   23.49 MB   1.41
  enp45s0u1u1  50016    UDP  184.82 k   65.19 k   3.61   35.90 MB   12.32 MB   0.84
  enp45s0u1u1  50059    UDP   33.80 k    3.45 k   0.54   29.29 MB  487.12 kB   0.52
  enp45s0u1u1    993    TCP    9.44 k    8.62 k   0.26    3.97 MB    1.26 MB   0.09
  enp45s0u1u1      0    UDP    1.00      2.08 k   0.03    1.47 kB    2.78 MB   0.05
  enp45s0u1u1  50050    UDP    1.44 k    2.26 k   0.05  173.56 kB    1.05 MB   0.02
                                  ...       ...               ...        ...       
                               4.29 M    2.63 M           4.21 GB    1.43 GB       
                                                                            
      Totals:                            6.92 M                      5.64 GB  

Timespan / Interface : [2022-08-18 11:31:19, 2022-08-18 17:56:19] / virbr0,lo,enp45s0u1u1
```
vs.
```
└─ $ ▶ goQuery -i any  -d /tmp/goprobe_db_pcap2 -f -10000d dport,proto -n 10

                              packets   packets             bytes      bytes       
        iface  dport  proto        in       out      %         in        out      %
  enp45s0u1u1    443    UDP    2.32 M  265.91 k  37.35    2.76 GB   25.99 MB  49.41
  enp45s0u1u1   3478    UDP  890.49 k    1.42 M  33.42  666.36 MB  913.75 MB  27.38
  enp45s0u1u1  50024    UDP  642.15 k  700.42 k  19.41  497.12 MB  435.56 MB  16.16
  enp45s0u1u1    443    TCP  114.95 k   96.88 k   3.06  192.06 MB   47.86 MB   4.16
  enp45s0u1u1   4500    UDP   89.55 k   63.22 k   2.21   58.12 MB   23.49 MB   1.41
  enp45s0u1u1  50016    UDP  184.79 k   65.22 k   3.61   35.90 MB   12.33 MB   0.84
  enp45s0u1u1  50059    UDP   30.21 k    3.02 k   0.48   26.54 MB  425.87 kB   0.47
  enp45s0u1u1    993    TCP    9.44 k    8.62 k   0.26    3.97 MB    1.26 MB   0.09
  enp45s0u1u1      0    UDP    1.00      2.08 k   0.03    1.47 kB    2.78 MB   0.05
  enp45s0u1u1  50050    UDP    1.44 k    2.26 k   0.05  173.67 kB    1.05 MB   0.02
                                  ...       ...               ...        ...       
                               4.29 M    2.63 M           4.21 GB    1.43 GB       
                                                                            
      Totals:                            6.92 M                      5.64 GB  

Timespan / Interface : [2022-08-18 11:31:19, 2022-08-18 17:56:19] / enp45s0u1u1,virbr0,lo
```
As can be seen, the overall sum of tracked packets is basically identical, but the two tuples for ports `3478` and `50024` show significant leakage. Most likely at some point in time flow expiry differed slightly (because of raciness), causing packets direction to come out differently in the next time frame.

**Comment by fako1024** [comment]
I'll try to narrow it down more, looking at the raw flows, but it'll require some magic to the output because the raw tables are not deterministic due to the 5min time resolution.

**Comment by els0r** [comment]
Great find! I will look into the code in more detail as well to get to the bottom of this.

An idea for debugging: since we don't really care about query performance while trying to reproduce the issue, goProbe could be modified to skip source port aggregation and store the source port in `sport.gpf`.

This would be a significant change in `goDB`, but might be worth it after all as it would not only guide us to the root of the problem, but also give us estimates just how _expensive_ the additional source port attribute is. Expensive along the lines of

- size of DB (% increase)
- query execution times
- memory footprint of goQuery (goProbe's will be nearly identical as it stores the sport internally)

**Comment by fako1024** [comment]
Fair point, could be doable, but at this point I think on my side I'd be ok with selective logging (After all, I know exactly on which connection / port it happens and can reproduce it basically every day as soon as I'm logged in to the company CVPN :-P ). But it's definitively something to keep in mind!

**Comment by fako1024** [comment]
Just a random idea off the top of my head: We could "invest" a little memory and persist a cache of all established UDP directions (based on the hash) beyond flow expiry, that way we wouldn't lose the information. In that case we could even assume that the first UDP packet seen for a UDP connection defines the direction (which right now we can't due to the expiry of information). Not perfect, but maybe something we could consider...

**Comment by fako1024** [comment]
This particular case could be "solved" as quick-fix by adding the STUN UDP port (3478) as special port to the heuristics (probably a _destination_ rather than a _source_ in >99% of the cases) as it's a quite common thing in these days in the context of Teams / Zoom etc.

**Comment by els0r** [comment]
Good point. Coming to think of it, maybe it's not even prudent to shoehorn TCP and UDP connections into the same logic. We could introduce two different Classifiers and (or actually: three) run them depending on the IP Protocol

- default
- tcp
- udp

This may make things slightly more robust while being more readable/understandable at the same time.

As for `>20000`, it's too long ago ... 😁

**Comment by els0r** [comment]
> Just a random idea off the top of my head: We could "invest" a little memory and persist a cache of all established UDP directions (based on the hash) beyond flow expiry, that way we wouldn't lose the information. In that case we could even assume that the first UDP packet seen for a UDP connection defines the direction (which right now we can't due to the expiry of information). Not perfect, but maybe something we could consider...

I'm not sure I fully understand it yet. Does this track long-running UDP flows which (at the boundary of a 5 minute write-out) the direction heuristic would have to be ran again?

Does this data structure expire, and if so, at what time? Or would it literally track every `(sip, dip, dport, proto:udp)` tuple and it's determined direction ever encountered by the system running `goProbe`? Just thinking of applications such as torrents, this would get very large very quickly.

**Comment by fako1024** [comment]
There's also some more, quite obviously incorrect detection failures. When querying an existing database I got several hits for ICMP flows with a non-zero dport:
```
   77.56.97.7        8.8.8.8   5405   ICMP  102.74 k  103.22 k  21.27    9.60 MB    9.64 MB  20.07
   77.56.97.7        8.8.4.4   5405   ICMP   58.76 k   58.96 k  12.16    5.49 MB    5.51 MB  11.47
   77.56.97.7    209.244.0.3   5405   ICMP   55.98 k   56.26 k  11.59    5.23 MB    5.26 MB  10.94
   77.56.97.7    209.244.0.4   5405   ICMP   56.01 k   56.16 k  11.59    5.23 MB    5.25 MB  10.93
  45.35.11.50     77.56.97.7   5405   ICMP   21.14 k    0.00     2.18    2.82 MB    0.00  B   2.94
      8.8.8.8     77.56.97.7   5405   ICMP   12.45 k   12.48 k   2.57    1.16 MB    1.17 MB   2.43
   77.56.97.7  144.76.51.228   5405   ICMP    0.00     13.70 k   1.42    0.00  B    2.04 MB   2.12
  209.244.0.4     77.56.97.7   5405   ICMP   10.18 k   10.18 k   2.10  973.78 kB  974.16 kB   1.98
      8.8.4.4     77.56.97.7   5405   ICMP   10.11 k   10.12 k   2.09  967.37 kB  968.42 kB   1.97
  209.244.0.3     77.56.97.7   5405   ICMP    6.75 k    6.76 k   1.40  645.90 kB  646.95 kB   1.32
```
First I thought it was an issue of the recent changes and the protocol got incorrectly translated, but the results are the same when performing the same query on the legacy format DB, so I'm confident that it's an intrinsic issue, not a DB / data format one.

**Comment by els0r** [comment]
We used to have ICMP in the detection heuristic and could leverage the `echo request` and `echo reply` flags to set the direction properly.

And we should set the sport & dport to `0x0000` for protocol ICMP

**Comment by fako1024** [comment]
Mh, indeed I remember that, but I also didn't remove it AFAIK. :thinking: - As for the protocol, IMHO it's not about _setting_ that to zero, it's more about: Why is it tracked as non-zero to begin with. It's been a while, but as far as I remember ICMP layers don't carry port information to begin with, so I'm wondering if that might even be a bug in gopacket... I'll run some tests and fir up a first shot at something new...

**Comment by els0r** [comment]
Yeah. ICMP doesn't carry port information, so I'm also wondering what those mysterious 4505 could be. Maybe some other parts of the ICMP protocol layer that are mistakenly put into the port slices?

Btw., I removed the ICMP handling since it was commented out for ages. We did that because we originally excluded ICMP in the BPF filter, thus not requiring specific handling in code. I'll dig it up and add it again.

**Comment by fako1024** [comment]
Port 5405 is used on that particular machine for a (UDP) OpenVPN tunnel to a fixed endpoint / machine outside my network, so there definitively is something fishy... But yeah, I'll try to figure it out, there's quite a lot of those, shouldn't be too hard to reproduce.
If you can't find the ICMP stuff quickly, don't bother, I'll just rewrite that passage anyway, figuring out that particular byte can't be that hard... :-P

**Comment by fako1024** [comment]
Belay that, I just missed it, it's still in the code:
```
	/* disregarded until further notice
	   if packet.ICMPtypeCode != 0xffff {
	       // try ICMP codes if protocol field matches. Echo request and reply can shed
	       // light on the situation. Other codes may help too
	       switch packet.ICMPtypeCode {
	       case 0x0000: // echo reply
	           return DirectionRemains
	       case 0x0800: // echo request
	           return DirectionReverts
	       }

	   }
	*/
```

**Comment by fako1024** [comment]
OK, there were a lot of issues in the current logic, e.g.:

- The TCP flag handling was plain wrong (e.g. FIN / FIN-ACK packets cannot be used to determine the direction, as _either_ side can terminate the session from their end. There is no way to know which one it is)
- The source port exception logic for port `53` wasn't working correctly, it actually skipped the source port for basically all UDP traffic because there was a `protocol == TCP` check for the rule (which is the opposite of what we want)
- The whole ephemeral / non ephemeral port logic was incomplete and too much focused on "special ports"
- ICMP packets and IPv6 were not handled completely / correctly / at all

I've completely rewritten the classification structure and will run it overnight on my router. So far the results are exactly what I expect them to be.

**Comment by els0r** [comment]
Fantastic. Thanks for revisiting this part!

**Comment by fako1024** [comment]
@fako1024 

- [ ] Note to self: Add UDP/443 for port merging logic:
